### PR TITLE
Add new features to SDE and transcribe speech

### PR DESCRIPTION
**Pull request by Jorjeous**
Signed-off-by: George Zelenfroind <gzelenfroind@nvidia.com>

# What does this PR do ?
Adds feature of comparing to models to SDE 
And adds new way of transcribing audio - add new transcripts to existing JSON file 

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]
asr

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below
* in readme

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tools/speech_data_explorer/data_explorer.py [hunk 1]
```diff
@@ -1,846 +1,1044 @@
...
-        ]
-    )
-    fig.update_layout(
-        showlegend=False, margin=dict(l=0, r=0, t=0, b=0, pad=0), height=200, yaxis={'title_text': '#words'}
-    )
-
-    return fig
-
-
-def absolute_audio_filepath(audio_filepath, audio_base_path):
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:330]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 2]
```diff
@@ -1,846 +1,1044 @@
...
-                    num_chars = 1e-9
-                data[-1]['WER'] = round(word_dist / num_words * 100.0, 2)
-                data[-1]['CER'] = round(char_dist / num_chars * 100.0, 2)
-                data[-1]['WMR'] = round(measures['hits'] / num_words * 100.0, 2)
-                data[-1]['I'] = measures['insertions']
-                data[-1]['D'] = measures['deletions']
-                data[-1]['D-I'] = measures['deletions'] - measures['insertions']
-
-            if estimate_audio:
-                filepath = absolute_audio_filepath(item['audio_filepath'], data_filename)
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:245]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 3]
```diff
@@ -1,846 +1,1044 @@
...
-        pickle_filename += '_' + timestamp + '.pkl'
-        if os.path.exists(pickle_filename):
-            with open(pickle_filename, 'rb') as f:
-                data, wer, cer, wmr, mwa, num_hours, vocabulary_data, alphabet, metrics_available = pickle.load(f)
-            if vocab is not None:
-                for item in vocabulary_data:
-                    item['OOV'] = item['word'] not in vocabulary_ext
-            if estimate_audio:
-                for item in data:
-                    filepath = absolute_audio_filepath(item['audio_filepath'], audio_base_path)
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:159]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 4]
```diff
@@ -1,846 +1,1044 @@
...
-    freqband = 0
-    for idx in range(len(power_spectrum) - 1, -1, -1):
-        if power_spectrum[idx] > threshold:
-            freqband = idx / n_fft * sr
-            break
-    return freqband
-
-
-# load data from JSON manifest file
-def load_data(data_filename, disable_caching=False, estimate_audio=False, vocab=None, audio_base_path=None):
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:131]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 5]
```diff
@@ -1,846 +1,1044 @@
...
-    parser.add_argument(
-        '--audio-base-path',
-        default=None,
-        type=str,
-        help='A base path for the relative paths in manifest. It defaults to manifest path.',
-    )
-    parser.add_argument('--debug', '-d', action='store_true', help='enable debug mode')
-    args = parser.parse_args()
-
-    # assume audio_filepath is relative to the directory where the manifest is stored
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:105]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 6]
```diff
@@ -1,846 +1,1044 @@
...
-    parser.add_argument(
-        '--disable-caching-metrics', action='store_true', help='disable caching metrics for errors analysis'
-    )
-    parser.add_argument(
-        '--estimate-audio-metrics',
-        '-a',
-        action='store_true',
-        help='estimate frequency bandwidth and signal level of audio recordings',
-    )
-    parser.add_argument(
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:96]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 7]
```diff
@@ -1,846 +1,1044 @@
...
-import datetime
-import difflib
-import io
-import json
-import math
-import operator
-import os
-import pickle
-from collections import defaultdict
-from os.path import expanduser
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:27]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 8]
```diff
@@ -1,846 +1,1044 @@
...
-
-    return filename
-
-
-# parse the CLI arguments
-args = parse_args()
-
-print('Loading data...')
-data, wer, cer, wmr, mwa, num_hours, vocabulary, alphabet, metrics_available = load_data(
-    args.manifest, args.disable_caching_metrics, args.estimate_audio_metrics, args.vocab, args.audio_base_path
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:355]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 9]
```diff
@@ -1,846 +1,1044 @@
...
+)
+def update_datatable(page_current, sort_by, filter_query):
+    data_view = data
+    filtering_expressions = filter_query.split(' && ')
+    for filter_part in filtering_expressions:
+        col_name, op, filter_value = split_filter_part(filter_part)
+
+        if op in ('eq', 'ne', 'lt', 'le', 'gt', 'ge'):
+            data_view = [x for x in data_view if getattr(operator, op)(x[col_name], filter_value)]
+        elif op == 'condtains':
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:882]: Please fix the introduced typo: `condtains` => `contains`

File: tools/speech_data_explorer/data_explorer.py [hunk 10]
```diff
@@ -1,846 +1,1044 @@
...
-    return diff_html
-
-
-@app.callback(Output('signal-graph', 'figure'), [Input('datatable', 'selected_rows'), Input('datatable', 'data')])
-def plot_signal(idx, data):
-    if len(idx) == 0:
-        raise PreventUpdate
-    figs = make_subplots(rows=2, cols=1, subplot_titles=('Waveform', 'Spectrogram'))
-    try:
-        filename = absolute_audio_filepath(data[idx[0]]['audio_filepath'], args.audio_base_path)
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:776]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 11]
```diff
@@ -1,846 +1,1044 @@
...
-
-    return figs
-
-
-@app.callback(Output('player', 'src'), [Input('datatable', 'selected_rows'), Input('datatable', 'data')])
-def update_player(idx, data):
-    if len(idx) == 0:
-        raise PreventUpdate
-    try:
-        filename = absolute_audio_filepath(data[idx[0]]['audio_filepath'], args.audio_base_path)
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:828]: Please integrate Ante's updates back (to support audio_base_path)

File: tools/speech_data_explorer/data_explorer.py [hunk 12]
```diff
@@ -1,846 +1,1044 @@
...
-            ),
-            row=2,
-            col=1,
-        )
-        figs.update_layout({'margin': dict(l=0, r=0, t=20, b=0, pad=0), 'height': 500})
-        figs.update_xaxes(title_text='Time, s', row=1, col=1)
-        figs.update_yaxes(title_text='Amplitude', row=1, col=1)
-        figs.update_xaxes(title_text='Time, s', row=2, col=1)
-        figs.update_yaxes(title_text='Frequency, kHz', row=2, col=1)
-    except Exception as ex:
```
Comment (vsl9) [inline tools/speech_data_explorer/data_explorer.py:817]: Please bring back the Exception handler (to log exception type)

**Review by vsl9 — CHANGES_REQUESTED** [review]

File: tools/speech_data_explorer/data_explorer.py (hunk 1 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:330]: Done

File: tools/speech_data_explorer/data_explorer.py (hunk 2 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:245]: done

File: tools/speech_data_explorer/data_explorer.py (hunk 4 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:131]: done

File: tools/speech_data_explorer/data_explorer.py (hunk 5 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:105]: done

File: tools/speech_data_explorer/data_explorer.py (hunk 7 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:27]: done

File: tools/speech_data_explorer/data_explorer.py (hunk 8 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:355]: done

File: tools/speech_data_explorer/data_explorer.py (hunk 9 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:882]: done

File: tools/speech_data_explorer/data_explorer.py (hunk 11 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:828]: done

File: tools/speech_data_explorer/data_explorer.py (hunk 12 above)
Comment (Jorjeous) [reply tools/speech_data_explorer/data_explorer.py:817]: done

**Comment by ericharper** [comment]
@Jorjeous can you please run `python setup.py style --fix` to fix code formatting?

**Comment by Jorjeous** [comment]
Will add better user interface)

**Pull request closed without merging**